RecoPixelVertexing/PixelVertexFinding/src/gpuClusterTracks.h

```diff
@@ -0,0 +1,243 @@
+#ifndef RecoPixelVertexing_PixelVertexFinding_clusterTracks_H
+#define RecoPixelVertexing_PixelVertexFinding_clusterTracks_H
+
+#include<cstdint>
+#include<cmath>
+#include<cassert>
+
+#include "HeterogeneousCore/CUDAUtilities/interface/HistoContainer.h"
+
+#include "gpuVertexFinder.h"
+
+namespace gpuVertexFinder {
+  
+  // this algo does not really scale as it works in a single block...
+  // enough for <10K tracks we have
+  __global__ 
+  void clusterTracks(int nt,
+		     OnGPU * pdata,
+		     int minT,  // min number of neighbours to be "core"
+		     float eps, // max absolute distance to cluster
+		     float errmax, // max error to be "seed"
+		     float chi2max   // max normalized distance to cluster
+		     )  {
+
+    constexpr bool verbose = false;
+    
+    auto er2mx = errmax*errmax;
+    
+    auto & data = *pdata;
+    float const * zt = data.zt;
+    float const * ezt2 = data.ezt2;
+    float * zv = data.zv;
+    float * wv = data.wv;
+    float * chi2 = data.chi2;
+    uint32_t & nv = *data.nv;
+    
+    int8_t  * izt = data.izt;
+    int32_t * nn = data.nn;
+    int32_t * iv = data.iv;
+    
+    assert(pdata);
+    assert(zt);
+    
+    __shared__ HistoContainer<int8_t,8,5,8,uint16_t> hist;
+    
+    //  if(0==threadIdx.x) printf("params %d %f\n",minT,eps);    
+    //  if(0==threadIdx.x) printf("booked hist with %d bins, size %d for %d tracks\n",hist.nbins(),hist.binSize(),nt);
+    
+    // zero hist
+    hist.nspills = 0;
+    for (auto k = threadIdx.x; k<hist.nbins(); k+=blockDim.x) hist.n[k]=0;
+    __syncthreads();
+
+    //  if(0==threadIdx.x) printf("histo zeroed\n");
+    
+    
+  // fill hist
+    for (int i = threadIdx.x; i < nt; i += blockDim.x) {
+      assert(i<64000);
+      int iz =  int(zt[i]*10.);
+      iz = std::max(iz,-127);
+      iz = std::min(iz,127);
+      izt[i]=iz;
+      hist.fill(int8_t(iz),uint16_t(i));
+      iv[i]=i;
+      nn[i]=0;
+    }
+    __syncthreads();
+    
+    //   if(0==threadIdx.x) printf("histo filled %d\n",hist.nspills);
+    if(0==threadIdx.x && hist.fullSpill()) printf("histo overflow\n");
+    
+    // count neighbours
+    for (int i = threadIdx.x; i < nt; i += blockDim.x) {
+      if (ezt2[i]>er2mx) continue;
+      auto loop = [&](int j) {
+        if (i==j) return;
+        auto dist = std::abs(zt[i]-zt[j]);
+        if (dist>eps) return;
+        if (dist*dist>chi2max*(ezt2[i]+ezt2[j])) return;
+        nn[i]++;
+      };
+      
+      int bs = hist.bin(izt[i]);
+      int be = std::min(int(hist.nbins()),bs+2);
+      bs = bs==0 ? 0 : bs-1;
+      assert(be>bs);
+      for (auto b=bs; b<be; ++b){
```

Comment:
do you think it would be possible to write an iterator-like interface for `hist` so that one can do this with a single (new style) loop ?

something like `for (auto pj: hist.range(bs, be))...` ?

(from the uses below I think it should always include the spills as well)

Reply:
I find iterator (with state) messy.
Maybe a function accepting a lambda (need to verify that cuda will really inline it)